oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/logout/OidcClientInitiatedServerLogoutSuccessHandler.java

```diff
@@ -0,0 +1,123 @@
+ */
+
+package org.springframework.security.oauth2.client.oidc.logout;
+
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+
+import reactor.core.publisher.Mono;
+
+import org.springframework.security.core.Authentication;
+import org.springframework.security.oauth2.client.authentication.OAuth2AuthenticationToken;
+import org.springframework.security.oauth2.client.registration.ClientRegistration;
+import org.springframework.security.oauth2.client.registration.ReactiveClientRegistrationRepository;
+import org.springframework.security.oauth2.core.oidc.user.OidcUser;
+import org.springframework.security.web.server.DefaultServerRedirectStrategy;
+import org.springframework.security.web.server.ServerRedirectStrategy;
+import org.springframework.security.web.server.WebFilterExchange;
+import org.springframework.security.web.server.authentication.logout.RedirectServerLogoutSuccessHandler;
+import org.springframework.security.web.server.authentication.logout.ServerLogoutSuccessHandler;
+import org.springframework.util.Assert;
+import org.springframework.web.util.UriComponentsBuilder;
+
+/**
+ * A reactive logout success handler for initiating OIDC logout through the user agent.
+ *
+ * @author Josh Cummings
+ * @since 5.2
+ * @see <a href="http://openid.net/specs/openid-connect-session-1_0.html#RPLogout">RP-Initiated Logout</a>
+ * @see {@link org.springframework.security.web.authentication.logout.LogoutSuccessHandler}
+ */
+public class OidcClientInitiatedServerLogoutSuccessHandler
+		implements ServerLogoutSuccessHandler {
+
+	private final ServerRedirectStrategy redirectStrategy = new DefaultServerRedirectStrategy();
+	private final ReactiveClientRegistrationRepository clientRegistrationRepository;
+
+	private URI logoutSuccessUrl = URI.create(RedirectServerLogoutSuccessHandler.DEFAULT_LOGOUT_SUCCESS_URL);
+	private URI postLogoutRedirectUri;
+
+	/**
+	 * Constructs an {@link OidcClientInitiatedServerLogoutSuccessHandler} with the provided parameters
+	 *
+	 * @param clientRegistrationRepository The {@link ReactiveClientRegistrationRepository} to use to derive
+	 * the end_session_endpoint value
+	 */
+	public OidcClientInitiatedServerLogoutSuccessHandler
+			(ReactiveClientRegistrationRepository clientRegistrationRepository) {
+
+		Assert.notNull(clientRegistrationRepository, "clientRegistrationRepository cannot be null");
+		this.clientRegistrationRepository = clientRegistrationRepository;
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	public Mono<Void> onLogoutSuccess(WebFilterExchange exchange, Authentication authentication) {
+		return Mono.just(authentication)
+				.filter(OAuth2AuthenticationToken.class::isInstance)
+				.filter(token -> authentication.getPrincipal() instanceof OidcUser)
+				.map(OAuth2AuthenticationToken.class::cast)
+				.flatMap(this::endSessionEndpoint)
+				.map(endSessionEndpoint -> endpointUri(endSessionEndpoint, authentication))
+				.defaultIfEmpty(this.logoutSuccessUrl)
```

Comment:
I think we could create a member (please consider renaming but illustrates point):

```java
private ServerLogoutSuccessHandler delegate = new RedirectServerLogoutSuccessHandler();

// add setter
```

You can then remove `this.logoutSuccessUrl` and change `.defaultIfEmpty(this.logoutSuccessUrl)` to

```java
.switchIfEmpty(delegate.onLogoutSuccess(exchange, authentication).then(Mono.empty()))
```